File: executor/protocol.go
package executor

import (
	"bytes"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"path/filepath"
	"regexp"
	"strings"
)

const (
	maxSpecBytes      = 1024 * 1024
	maxArgCount       = 256
	maxArgBytes       = 32 * 1024
	maxEnvCount       = 256
	maxEnvValueBytes  = 128 * 1024
	maxTimeoutMillis       = 24 * 60 * 60 * 1000
	maxCaptureBytes        = 16 * 1024 * 1024
	maxRunIDBytes          = 256
	maxObligationIDBytes   = 512
	maxRevisionBytes       = 256
	maxAuthorityCommitBytes = 256
	maxCapabilityBytes     = 4096
	maxExecutionGeneration = 9007199254740991
)

var envKeyPattern = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

type ProcessSpecV1 struct {
	Schema         string            `json:"schema"`
	Executable     string            `json:"executable"`
	Argv           []string          `json:"argv"`
	Cwd            string            `json:"cwd"`
	Env            map[string]string `json:"env"`
	TimeoutMillis  int64             `json:"timeout_ms"`
	StdoutMaxBytes int64             `json:"stdout_max_bytes"`
	StderrMaxBytes int64             `json:"stderr_max_bytes"`
}

type ComputationExecutionV1 struct {
	Schema                    string `json:"schema"`
	RunID                     string `json:"run_id"`
	ObligationID              string `json:"obligation_id"`
	ClaimedRevision           string `json:"claimed_revision"`
	ExecutionGeneration       int64  `json:"execution_generation"`
	ExecutionAuthorityCommit  string `json:"execution_authority_commit"`
	ExecutionCapability       string `json:"execution_capability"`
	ExecutionCapabilitySHA256 string `json:"execution_capability_sha256"`
	ExecutionSpecBase64       string `json:"execution_spec_base64"`
	ExecutionSpecSHA256       string `json:"execution_spec_sha256"`
}

type ExecutionIdentityV1 struct {
	RunID                    string `json:"run_id"`
	ExecutionGeneration      int64  `json:"execution_generation"`
	ExecutionAuthorityCommit string `json:"execution_authority_commit"`
}

type ExecutorCommandV1 struct {
	Schema    string
	Kind      string
	Execution *ComputationExecutionV1
	Identity  *ExecutionIdentityV1
	validated *validatedExecution
}

type ComputationAttemptEvidenceV1 struct {
	Schema                    string `json:"schema"`
	RunID                     string `json:"run_id"`
	ObligationID              string `json:"obligation_id"`
	ClaimedRevision           string `json:"claimed_revision"`
	ExecutionGeneration       int64  `json:"execution_generation"`
	ExecutionAuthorityCommit  string `json:"execution_authority_commit"`
	ExecutionCapabilitySHA256 string `json:"execution_capability_sha256"`
	ExecutionSpecSHA256       string `json:"execution_spec_sha256"`
	Outcome                   string `json:"outcome"`
	ExitCode                  *int   `json:"exit_code,omitempty"`
	Signal                    string `json:"signal,omitempty"`
	StdoutBase64              string `json:"stdout_base64,omitempty"`
	StdoutSHA256              string `json:"stdout_sha256"`
	StdoutTruncated           bool   `json:"stdout_truncated"`
	StderrBase64              string `json:"stderr_base64,omitempty"`
	StderrSHA256              string `json:"stderr_sha256"`
	StderrTruncated           bool   `json:"stderr_truncated"`
	Error                     string `json:"error,omitempty"`
}

type validatedExecution struct {
	Execution ComputationExecutionV1
	Spec      ProcessSpecV1
	SpecBytes []byte
}

func decodeStrict(data []byte, target any) error {
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(target); err != nil {
		return err
	}
	if err := decoder.Decode(&struct{}{}); !errors.Is(err, io.EOF) {
		if err == nil {
			return errors.New("multiple JSON values")
		}
		return err
	}
	return nil
}

func validateExactObjectKeys(
	data []byte,
	required []string,
	optional []string,
	name string,
) error {
	var object map[string]json.RawMessage
	if err := decodeStrict(data, &object); err != nil {
		return fmt.Errorf("invalid %s: %w", name, err)
	}
	allowed := make(map[string]struct{}, len(required)+len(optional))
	for _, key := range required {
		allowed[key] = struct{}{}
	}
	for _, key := range optional {
		allowed[key] = struct{}{}
	}
	for key := range object {
		if _, ok := allowed[key]; !ok {
			return fmt.Errorf("%s unknown field: %s", name, key)
		}
	}
	for _, key := range required {
		if _, ok := object[key]; !ok {
			return fmt.Errorf("%s missing field: %s", name, key)
		}
	}
	return nil
}

func validateString(value, name string, maxBytes int) error {
	if value == "" || strings.ContainsRune(value, 0) || len([]byte(value)) > maxBytes {
		return fmt.Errorf("%s invalid", name)
	}
	return nil
}

func validateIdentity(identity ExecutionIdentityV1) error {
	if err := validateString(identity.RunID, "run_id", maxRunIDBytes); err != nil {
		return err
	}
	if identity.ExecutionGeneration <= 0 || identity.ExecutionGeneration > maxExecutionGeneration {
		return errors.New("execution_generation invalid")
	}
	if err := validateString(identity.ExecutionAuthorityCommit, "execution_authority_commit", maxAuthorityCommitBytes); err != nil {
		return err
	}
	return nil
}

type executionIdentityKey struct {
	RunID                    string
	ExecutionGeneration      int64
	ExecutionAuthorityCommit string
}

func identityKey(identity ExecutionIdentityV1) executionIdentityKey {
	return executionIdentityKey{
		RunID:                    identity.RunID,
		ExecutionGeneration:      identity.ExecutionGeneration,
		ExecutionAuthorityCommit: identity.ExecutionAuthorityCommit,
	}
}

func identityFor(execution ComputationExecutionV1) ExecutionIdentityV1 {
	return ExecutionIdentityV1{
		RunID:                    execution.RunID,
		ExecutionGeneration:      execution.ExecutionGeneration,
		ExecutionAuthorityCommit: execution.ExecutionAuthorityCommit,
	}
}

func parseCommand(line []byte) (ExecutorCommandV1, error) {
	if err := validateExactObjectKeys(
		line,
		[]string{"schema", "kind"},
		[]string{"execution", "identity"},
		"executor command",
	); err != nil {
		return ExecutorCommandV1{}, err
	}
	var wire struct {
		Schema    string          `json:"schema"`
		Kind      string          `json:"kind"`
		Execution json.RawMessage `json:"execution,omitempty"`
		Identity  json.RawMessage `json:"identity,omitempty"`
	}
	if err := decodeStrict(line, &wire); err != nil {
		return ExecutorCommandV1{}, fmt.Errorf("invalid executor command: %w", err)
	}
	if wire.Schema != ExecutorCommandSchema {
		return ExecutorCommandV1{}, errors.New("executor command schema mismatch")
	}

	switch wire.Kind {
	case "execute":
		if len(wire.Execution) == 0 || len(wire.Identity) != 0 {
			return ExecutorCommandV1{}, errors.New("execute command shape invalid")
		}
		validated, err := validateExecutionBytes(wire.Execution)
		if err != nil {
			return ExecutorCommandV1{}, err
		}
		return ExecutorCommandV1{
			Schema:    wire.Schema,
			Kind:      wire.Kind,
			Execution: &validated.Execution,
			validated: &validated,
		}, nil
	case "cancel":
		if len(wire.Identity) == 0 || len(wire.Execution) != 0 {
			return ExecutorCommandV1{}, errors.New("cancel command shape invalid")
		}
		if err := validateExactObjectKeys(
			wire.Identity,
			[]string{"run_id", "execution_generation", "execution_authority_commit"},
			nil,
			"cancel identity",
		); err != nil {
			return ExecutorCommandV1{}, err
		}
		var identity ExecutionIdentityV1
		if err := decodeStrict(wire.Identity, &identity); err != nil {
			return ExecutorCommandV1{}, fmt.Errorf("invalid cancel identity: %w", err)
		}
		if err := validateIdentity(identity); err != nil {
			return ExecutorCommandV1{}, err
		}
		return ExecutorCommandV1{
			Schema:   wire.Schema,
			Kind:     wire.Kind,
			Identity: &identity,
		}, nil
	default:
		return ExecutorCommandV1{}, errors.New("executor command kind invalid")
	}
}

func validateExecutionBytes(data []byte) (validatedExecution, error) {
	if err := validateExactObjectKeys(
		data,
		[]string{
			"schema",
			"run_id",
			"obligation_id",
			"claimed_revision",
			"execution_generation",
			"execution_authority_commit",
			"execution_capability",
			"execution_capability_sha256",
			"execution_spec_base64",
			"execution_spec_sha256",
		},
		nil,
		"computation execution",
	); err != nil {
		return validatedExecution{}, err
	}
	var execution ComputationExecutionV1
	if err := decodeStrict(data, &execution); err != nil {
		return validatedExecution{}, fmt.Errorf("invalid computation execution: %w", err)
	}
	if execution.Schema != ComputationExecutionSchema {
		return validatedExecution{}, errors.New("computation execution schema mismatch")
	}
	for _, field := range []struct {
		name     string
		value    string
		maxBytes int
	}{
		{name: "run_id", value: execution.RunID, maxBytes: maxRunIDBytes},
		{name: "obligation_id", value: execution.ObligationID, maxBytes: maxObligationIDBytes},
		{name: "claimed_revision", value: execution.ClaimedRevision, maxBytes: maxRevisionBytes},
		{name: "execution_authority_commit", value: execution.ExecutionAuthorityCommit, maxBytes: maxAuthorityCommitBytes},
	} {
		if err := validateString(field.value, field.name, field.maxBytes); err != nil {
			return validatedExecution{}, err
		}
	}
	if execution.ExecutionGeneration <= 0 || execution.ExecutionGeneration > maxExecutionGeneration {
		return validatedExecution{}, errors.New("execution_generation invalid")
	}
	if err := validateString(execution.ExecutionCapability, "execution_capability", maxCapabilityBytes); err != nil {
		return validatedExecution{}, err
	}
	if len(execution.ExecutionCapabilitySHA256) != 64 {
		return validatedExecution{}, errors.New("execution_capability_sha256 invalid")
	}
	capabilityDigest := sha256.Sum256([]byte(execution.ExecutionCapability))
	if hex.EncodeToString(capabilityDigest[:]) != execution.ExecutionCapabilitySHA256 {
		return validatedExecution{}, errors.New("execution capability digest mismatch")
	}

	specBytes, err := base64.StdEncoding.DecodeString(execution.ExecutionSpecBase64)
	if err != nil || len(specBytes) == 0 || len(specBytes) > maxSpecBytes {
		return validatedExecution{}, errors.New("execution_spec_base64 invalid")
	}
	if base64.StdEncoding.EncodeToString(specBytes) != execution.ExecutionSpecBase64 {
		return validatedExecution{}, errors.New("execution_spec_base64 non-canonical")
	}
	specDigest := sha256.Sum256(specBytes)
	if execution.ExecutionSpecSHA256 != "sha256:"+hex.EncodeToString(specDigest[:]) {
		return validatedExecution{}, errors.New("execution spec digest mismatch")
	}

	spec, err := validateProcessSpecBytes(specBytes)
	if err != nil {
		return validatedExecution{}, err
	}
	return validatedExecution{
		Execution: execution,
		Spec:      spec,
		SpecBytes: specBytes,
	}, nil
}

func validateProcessSpecBytes(data []byte) (ProcessSpecV1, error) {
	if err := validateExactObjectKeys(
		data,
		[]string{
			"schema",
			"executable",
			"argv",
			"cwd",
			"env",
			"timeout_ms",
			"stdout_max_bytes",
			"stderr_max_bytes",
		},
		nil,
		"process spec",
	); err != nil {
		return ProcessSpecV1{}, err
	}
	var spec ProcessSpecV1
	if err := decodeStrict(data, &spec); err != nil {
		return ProcessSpecV1{}, fmt.Errorf("invalid process spec: %w", err)
	}
	if spec.Schema != ProcessSpecSchema {
		return ProcessSpecV1{}, errors.New("process spec schema mismatch")
	}
	if err := validateString(spec.Executable, "executable", 4096); err != nil {
		return ProcessSpecV1{}, err
	}
	if !filepath.IsAbs(spec.Executable) {
		return ProcessSpecV1{}, errors.New("executable must be absolute")
	}
	if len(spec.Argv) > maxArgCount {
		return ProcessSpecV1{}, errors.New("argv invalid")
	}
	for index, arg := range spec.Argv {
		if strings.ContainsRune(arg, 0) || len([]byte(arg)) > maxArgBytes {
			return ProcessSpecV1{}, fmt.Errorf("argv invalid at index %d", index)
		}
	}
	if err := validateString(spec.Cwd, "cwd", 4096); err != nil {
		return ProcessSpecV1{}, err
	}
	if filepath.IsAbs(spec.Cwd) {
		return ProcessSpecV1{}, errors.New("cwd must be relative")
	}
	clean := filepath.Clean(spec.Cwd)
	if clean == ".." || strings.HasPrefix(clean, ".."+string(filepath.Separator)) {
		return ProcessSpecV1{}, errors.New("cwd escapes workspace")
	}
	spec.Cwd = clean
	if len(spec.Env) > maxEnvCount {
		return ProcessSpecV1{}, errors.New("env too large")
	}
	for key, value := range spec.Env {
		if !envKeyPattern.MatchString(key) {
			return ProcessSpecV1{}, fmt.Errorf("env key invalid: %s", key)
		}
		if strings.ContainsRune(value, 0) || len([]byte(value)) > maxEnvValueBytes {
			return ProcessSpecV1{}, fmt.Errorf("env value invalid: %s", key)
		}
	}
	if spec.TimeoutMillis < 1 || spec.TimeoutMillis > maxTimeoutMillis {
		return ProcessSpecV1{}, errors.New("timeout_ms invalid")
	}
	if spec.StdoutMaxBytes < 0 || spec.StdoutMaxBytes > maxCaptureBytes {
		return ProcessSpecV1{}, errors.New("stdout_max_bytes invalid")
	}
	if spec.StderrMaxBytes < 0 || spec.StderrMaxBytes > maxCaptureBytes {
		return ProcessSpecV1{}, errors.New("stderr_max_bytes invalid")
	}
	return spec, nil
}

File: src/execution/protocol.ts
import { createHash } from 'node:crypto';
import { posix as path } from 'node:path';
import type { ExecutionPermit } from '../model.ts';

export const COMPUTATION_EXECUTION_SCHEMA='overcenter-computation-execution-v1' as const;
export const PROCESS_SPEC_SCHEMA='overcenter-process-spec-v1' as const;
export const COMPUTATION_EVIDENCE_SCHEMA='overcenter-computation-attempt-evidence-v1' as const;
export const EXECUTOR_COMMAND_SCHEMA='overcenter-executor-command-v1' as const;
export const EXECUTOR_HELLO_SCHEMA='overcenter-executor-hello-v1' as const;

const MAX_SPEC_BYTES=1024*1024;
const MAX_ARG_COUNT=256;
const MAX_ARG_BYTES=32*1024;
const MAX_ENV_COUNT=256;
const MAX_ENV_VALUE_BYTES=128*1024;
const MAX_TIMEOUT_MS=24*60*60*1000;
const MAX_CAPTURE_BYTES=16*1024*1024;

export interface ProcessSpecV1 {
  schema:typeof PROCESS_SPEC_SCHEMA;
  executable:string;
  argv:string[];
  cwd:string;
  env:Record<string,string>;
  timeout_ms:number;
  stdout_max_bytes:number;
  stderr_max_bytes:number;
}

export interface ComputationExecutionV1 {
  schema:typeof COMPUTATION_EXECUTION_SCHEMA;
  run_id:string;
  obligation_id:string;
  claimed_revision:string;
  execution_generation:number;
  execution_authority_commit:string;
  execution_capability:string;
  execution_capability_sha256:string;
  execution_spec_base64:string;
  execution_spec_sha256:string;
}

export interface ExecutionIdentityV1 {
  run_id:string;
  execution_generation:number;
  execution_authority_commit:string;
}

export interface ExecutorHelloV1 {
  schema:typeof EXECUTOR_HELLO_SCHEMA;
  execution_context_sha256:string;
  containment_id:string;
}

export type ExecutorCommandV1 =
  | {
      schema:typeof EXECUTOR_COMMAND_SCHEMA;
      kind:'execute';
      execution:ComputationExecutionV1;
    }
  | {
      schema:typeof EXECUTOR_COMMAND_SCHEMA;
      kind:'cancel';
      identity:ExecutionIdentityV1;
    };

export interface ComputationAttemptEvidenceV1 {
  schema:typeof COMPUTATION_EVIDENCE_SCHEMA;
  run_id:string;
  obligation_id:string;
  claimed_revision:string;
  execution_generation:number;
  execution_authority_commit:string;
  execution_capability_sha256:string;
  execution_spec_sha256:string;
  outcome:'completed'|'failed'|'cancelled';
  exit_code?:number;
  signal?:string;
  stdout_base64?:string;
  stdout_sha256:string;
  stdout_truncated:boolean;
  stderr_base64?:string;
  stderr_sha256:string;
  stderr_truncated:boolean;
  error?:string;
}

function sha256Hex(bytes:Uint8Array|string):string {
  return createHash('sha256').update(bytes).digest('hex');
}

function sha256Tagged(bytes:Uint8Array|string):string {
  return 'sha256:'+sha256Hex(bytes);
}

function assertPlainObject(value:unknown,name:string):asserts value is Record<string,unknown> {
  if (!value || typeof value!=='object' || Array.isArray(value)) {
    throw new Error(`${name.toUpperCase()}_INVALID`);
  }
}

function assertExactKeys(
  value:Record<string,unknown>,
  required:string[],
  optional:string[]=[],
  name='object',
):void {
  const allowed=new Set([...required,...optional]);
  for (const key of Object.keys(value)) {
    if (!allowed.has(key)) throw new Error(`${name.toUpperCase()}_UNKNOWN_FIELD:${key}`);
  }
  for (const key of required) {
    if (!(key in value)) throw new Error(`${name.toUpperCase()}_MISSING_FIELD:${key}`);
  }
}

function assertBoundedString(value:unknown,name:string,maxBytes=4096):asserts value is string {
  if (
    typeof value!=='string'
    || value.length===0
    || value.includes('\0')
    || Buffer.byteLength(value,'utf8')>maxBytes
  ) {
    throw new Error(`${name.toUpperCase()}_INVALID`);
  }
}

function assertSafeIntegerRange(value:unknown,name:string,min:number,max:number):asserts value is number {
  if (!Number.isSafeInteger(value) || (value as number)<min || (value as number)>max) {
    throw new Error(`${name.toUpperCase()}_INVALID`);
  }
}

function assertSha256Hex(value:unknown,name:string):asserts value is string {
  if (typeof value!=='string' || !/^[0-9a-f]{64}$/.test(value)) {
    throw new Error(`${name.toUpperCase()}_INVALID`);
  }
}

function assertSha256Tagged(value:unknown,name:string):asserts value is string {
  if (typeof value!=='string' || !/^sha256:[0-9a-f]{64}$/.test(value)) {
    throw new Error(`${name.toUpperCase()}_INVALID`);
  }
}

export function validateExecutorHello(value:unknown):ExecutorHelloV1 {
  assertPlainObject(value,'executor_hello');
  assertExactKeys(value,[
    'schema',
    'execution_context_sha256',
    'containment_id',
  ],[],'executor_hello');
  if (value.schema!==EXECUTOR_HELLO_SCHEMA) {
    throw new Error('EXECUTOR_HELLO_SCHEMA_MISMATCH');
  }
  assertSha256Tagged(value.execution_context_sha256,'execution_context_sha256');
  assertBoundedString(value.containment_id,'containment_id',512);
  return {
    schema:EXECUTOR_HELLO_SCHEMA,
    execution_context_sha256:value.execution_context_sha256,
    containment_id:value.containment_id,
  };
}

function decodeCanonicalBase64(
  value:unknown,
  maxBytes=MAX_SPEC_BYTES,
  errorCode='EXECUTION_SPEC_BASE64_INVALID',
):Buffer {
  if (typeof value!=='string' || value.length===0) {
    throw new Error(errorCode);
  }
  const bytes=Buffer.from(value,'base64');
  if (bytes.length===0 || bytes.length>maxBytes || bytes.toString('base64')!==value) {
    throw new Error(errorCode);
  }
  return bytes;
}

export function validateProcessSpec(value:unknown):ProcessSpecV1 {
  assertPlainObject(value,'process_spec');
  assertExactKeys(value,[
    'schema',
    'executable',
    'argv',
    'cwd',
    'env',
    'timeout_ms',
    'stdout_max_bytes',
    'stderr_max_bytes',
  ],[],'process_spec');

  if (value.schema!==PROCESS_SPEC_SCHEMA) throw new Error('PROCESS_SPEC_SCHEMA_MISMATCH');
  assertBoundedString(value.executable,'executable',4096);
  if (!path.isAbsolute(value.executable)) throw new Error('EXECUTABLE_MUST_BE_ABSOLUTE');

  if (!Array.isArray(value.argv) || value.argv.length>MAX_ARG_COUNT) {
    throw new Error('ARGV_INVALID');
  }
  const argv=value.argv.map((arg,index)=>{
    if (typeof arg!=='string' || arg.includes('\0') || Buffer.byteLength(arg,'utf8')>MAX_ARG_BYTES) {
      throw new Error(`ARGV_INVALID:${index}`);
    }
    return arg;
  });

  assertBoundedString(value.cwd,'cwd',4096);
  if (path.isAbsolute(value.cwd)) throw new Error('CWD_MUST_BE_RELATIVE');
  const clean=path.normalize(value.cwd);
  if (clean==='..' || clean.startsWith('../')) throw new Error('CWD_ESCAPES_WORKSPACE');

  assertPlainObject(value.env,'env');
  const entries=Object.entries(value.env);
  if (entries.length>MAX_ENV_COUNT) throw new Error('ENV_TOO_LARGE');
  const env:Record<string,string>={};
  for (const [key,raw] of entries) {
    if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(key)) throw new Error(`ENV_KEY_INVALID:${key}`);
    if (typeof raw!=='string' || raw.includes('\0') || Buffer.byteLength(raw,'utf8')>MAX_ENV_VALUE_BYTES) {
      throw new Error(`ENV_VALUE_INVALID:${key}`);
    }
    env[key]=raw;
  }

  assertSafeIntegerRange(value.timeout_ms,'timeout_ms',1,MAX_TIMEOUT_MS);
  assertSafeIntegerRange(value.stdout_max_bytes,'stdout_max_bytes',0,MAX_CAPTURE_BYTES);
  assertSafeIntegerRange(value.stderr_max_bytes,'stderr_max_bytes',0,MAX_CAPTURE_BYTES);

  return {
    schema:PROCESS_SPEC_SCHEMA,
    executable:value.executable,
    argv,
    cwd:clean,
    env,
    timeout_ms:value.timeout_ms,
    stdout_max_bytes:value.stdout_max_bytes,
    stderr_max_bytes:value.stderr_max_bytes,
  };
}

export function encodeProcessSpec(spec:ProcessSpecV1):{
  bytes:Buffer;
  base64:string;
  sha256:string;
} {
  const validated=validateProcessSpec(spec);
  const bytes=Buffer.from(JSON.stringify(validated),'utf8');
  if (bytes.length>MAX_SPEC_BYTES) throw new Error('EXECUTION_SPEC_TOO_LARGE');
  return {
    bytes,
    base64:bytes.toString('base64'),
    sha256:sha256Tagged(bytes),
  };
}

export function decodeProcessSpec(execution:ComputationExecutionV1):ProcessSpecV1 {
  validateComputationExecution(execution);
  const bytes=decodeCanonicalBase64(execution.execution_spec_base64);
  let parsed:unknown;
  try {
    parsed=JSON.parse(bytes.toString('utf8'));
  } catch {
    throw new Error('EXECUTION_SPEC_JSON_INVALID');
  }
  return validateProcessSpec(parsed);
}

export function validateComputationExecution(value:unknown):ComputationExecutionV1 {
  assertPlainObject(value,'computation_execution');
  assertExactKeys(value,[
    'schema',
    'run_id',
    'obligation_id',
    'claimed_revision',
    'execution_generation',
    'execution_authority_commit',
    'execution_capability',
    'execution_capability_sha256',
    'execution_spec_base64',
    'execution_spec_sha256',
  ],[],'computation_execution');

  if (value.schema!==COMPUTATION_EXECUTION_SCHEMA) throw new Error('COMPUTATION_EXECUTION_SCHEMA_MISMATCH');
  assertBoundedString(value.run_id,'run_id',256);
  assertBoundedString(value.obligation_id,'obligation_id',512);
  assertBoundedString(value.claimed_revision,'claimed_revision',256);
  assertSafeIntegerRange(value.execution_generation,'execution_generation',1,Number.MAX_SAFE_INTEGER);
  assertBoundedString(value.execution_authority_commit,'execution_authority_commit',256);
  assertBoundedString(value.execution_capability,'execution_capability',4096);
  assertSha256Hex(value.execution_capability_sha256,'execution_capability_sha256');
  if (sha256Hex(value.execution_capability)!==value.execution_capability_sha256) {
    throw new Error('EXECUTION_CAPABILITY_DIGEST_MISMATCH');
  }

  const specBytes=decodeCanonicalBase64(value.execution_spec_base64);
  assertSha256Tagged(value.execution_spec_sha256,'execution_spec_sha256');
  if (sha256Tagged(specBytes)!==value.execution_spec_sha256) {
    throw new Error('EXECUTION_SPEC_DIGEST_MISMATCH');
  }

  return {
    schema:COMPUTATION_EXECUTION_SCHEMA,
    run_id:value.run_id,
    obligation_id:value.obligation_id,
    claimed_revision:value.claimed_revision,
    execution_generation:value.execution_generation,
    execution_authority_commit:value.execution_authority_commit,
    execution_capability:value.execution_capability,
    execution_capability_sha256:value.execution_capability_sha256,
    execution_spec_base64:value.execution_spec_base64,
    execution_spec_sha256:value.execution_spec_sha256,
  };
}

export function computationExecution(
  permit:ExecutionPermit,
  spec:ProcessSpecV1,
):ComputationExecutionV1 {
  const encoded=encodeProcessSpec(spec);
  return validateComputationExecution({
    schema:COMPUTATION_EXECUTION_SCHEMA,
    run_id:permit.id,
    obligation_id:permit.obligation_id,
    claimed_revision:permit.claimed_revision,
    execution_generation:permit.execution_generation,
    execution_authority_commit:permit.execution_authority_commit,
    execution_capability:permit.execution_capability,
    execution_capability_sha256:permit.execution_capability_sha256,
    execution_spec_base64:encoded.base64,
    execution_spec_sha256:encoded.sha256,
  });
}

export function executionIdentity(execution:ComputationExecutionV1):ExecutionIdentityV1 {
  return {
    run_id:execution.run_id,
    execution_generation:execution.execution_generation,
    execution_authority_commit:execution.execution_authority_commit,
  };
}

export function executionIdentityKey(identity:ExecutionIdentityV1):string {
  return JSON.stringify([
    identity.run_id,
    identity.execution_generation,
    identity.execution_authority_commit,
  ]);
}


export function validateComputationEvidence(value:unknown):ComputationAttemptEvidenceV1 {
  assertPlainObject(value,'computation_evidence');
  assertExactKeys(value,[
    'schema',
    'run_id',
    'obligation_id',
    'claimed_revision',
    'execution_generation',
    'execution_authority_commit',
    'execution_capability_sha256',
    'execution_spec_sha256',
    'outcome',
    'stdout_sha256',
    'stdout_truncated',
    'stderr_sha256',
    'stderr_truncated',
  ],[
    'exit_code',
    'signal',
    'stdout_base64',
    'stderr_base64',
    'error',
  ],'computation_evidence');

  if (value.schema!==COMPUTATION_EVIDENCE_SCHEMA) {
    throw new Error('COMPUTATION_EVIDENCE_SCHEMA_MISMATCH');
  }
  assertBoundedString(value.run_id,'run_id',256);
  assertBoundedString(value.obligation_id,'obligation_id',512);
  assertBoundedString(value.claimed_revision,'claimed_revision',256);
  assertSafeIntegerRange(value.execution_generation,'execution_generation',1,Number.MAX_SAFE_INTEGER);
  assertBoundedString(value.execution_authority_commit,'execution_authority_commit',256);
  assertSha256Hex(value.execution_capability_sha256,'execution_capability_sha256');
  assertSha256Tagged(value.execution_spec_sha256,'execution_spec_sha256');
  if (!['completed','failed','cancelled'].includes(String(value.outcome))) {
    throw new Error('COMPUTATION_EVIDENCE_OUTCOME_INVALID');
  }
  if (value.exit_code!==undefined) {
    assertSafeIntegerRange(value.exit_code,'exit_code',0,255);
  }
  if (value.signal!==undefined && (typeof value.signal!=='string' || value.signal.length===0)) {
    throw new Error('COMPUTATION_EVIDENCE_SIGNAL_INVALID');
  }
  if (value.stdout_base64!==undefined) {
    decodeCanonicalBase64(
      value.stdout_base64,
      MAX_CAPTURE_BYTES,
      'COMPUTATION_EVIDENCE_STDOUT_BASE64_INVALID',
    );
  }
  if (value.stderr_base64!==undefined) {
    decodeCanonicalBase64(
      value.stderr_base64,
      MAX_CAPTURE_BYTES,
      'COMPUTATION_EVIDENCE_STDERR_BASE64_INVALID',
    );
  }
  assertSha256Tagged(value.stdout_sha256,'stdout_sha256');
  assertSha256Tagged(value.stderr_sha256,'stderr_sha256');
  if (typeof value.stdout_truncated!=='boolean' || typeof value.stderr_truncated!=='boolean') {
    throw new Error('COMPUTATION_EVIDENCE_TRUNCATION_INVALID');
  }
  if (value.error!==undefined && typeof value.error!=='string') {
    throw new Error('COMPUTATION_EVIDENCE_ERROR_INVALID');
  }

  return value as unknown as ComputationAttemptEvidenceV1;
}

function assertCapturedStreamForSpec(
  base64:string|undefined,
  digest:string,
  truncated:boolean,
  maxBytes:number,
  stream:'stdout'|'stderr',
):void {
  const captured=base64===undefined
    ? Buffer.alloc(0)
    : decodeCanonicalBase64(
        base64,
        maxBytes,
        `COMPUTATION_EVIDENCE_${stream.toUpperCase()}_BASE64_INVALID`,
      );
  if (truncated && captured.length!==maxBytes) {
    throw new Error(`COMPUTATION_EVIDENCE_${stream.toUpperCase()}_TRUNCATION_MISMATCH`);
  }
  if (!truncated && sha256Tagged(captured)!==digest) {
    throw new Error(`COMPUTATION_EVIDENCE_${stream.toUpperCase()}_DIGEST_MISMATCH`);
  }
}

export function assertComputationEvidenceFor(
  evidence:ComputationAttemptEvidenceV1,
  execution:ComputationExecutionV1,
):void {
  validateComputationEvidence(evidence);
  if (evidence.schema!==COMPUTATION_EVIDENCE_SCHEMA) {
    throw new Error('COMPUTATION_EVIDENCE_SCHEMA_MISMATCH');
  }
  if (
    evidence.run_id!==execution.run_id
    || evidence.obligation_id!==execution.obligation_id
    || evidence.claimed_revision!==execution.claimed_revision
    || evidence.execution_generation!==execution.execution_generation
    || evidence.execution_authority_commit!==execution.execution_authority_commit
    || evidence.execution_capability_sha256!==execution.execution_capability_sha256
  ) {
    throw new Error('COMPUTATION_EVIDENCE_AUTHORITY_MISMATCH');
  }
  if (evidence.execution_spec_sha256!==execution.execution_spec_sha256) {
    throw new Error('COMPUTATION_EVIDENCE_SPEC_MISMATCH');
  }

  const spec=decodeProcessSpec(execution);
  assertCapturedStreamForSpec(
    evidence.stdout_base64,
    evidence.stdout_sha256,
    evidence.stdout_truncated,
    spec.stdout_max_bytes,
    'stdout',
  );
  assertCapturedStreamForSpec(
    evidence.stderr_base64,
    evidence.stderr_sha256,
    evidence.stderr_truncated,
    spec.stderr_max_bytes,
    'stderr',
  );
}
